plasmapy/examples/plot_particle_stepper.py

```diff
@@ -2,78 +2,113 @@
+def magnetic_field(r):
+    return u.Quantity([4, 0, 0], u.T)
+
 
-B0 = 4 * u.T
-plasma.magnetic_field[0, :, :, :] = np.ones((10, 10, 10)) * B0
+# precomputed for efficiency
+E_unit = u.V / u.m
 
-E0 = 2 * u.V / u.m
-plasma.electric_field[1, :, :, :] = np.ones((10, 10, 10)) * E0
+
+def electric_field(r):
+    return u.Quantity([0, 2, 0], E_unit)
+
+
+plasma = AnalyticalFields(magnetic_field, electric_field)
 
 ############################################################
-# Calculate the timestep. We'll take one proton `p`, take its gyrofrequency, invert that
+# We'll now calculate the timestep. We'll take one proton `p`,
+# take its gyrofrequency, invert that
 # to get to the gyroperiod, and resolve that into 10 steps for higher accuracy.
 
-freq = gyrofrequency(B0, 'p').to(u.Hz, equivalencies=u.dimensionless_angles())
-gyroperiod = (1/freq).to(u.s)
+freq = gyrofrequency(4 * u.T, "p").to(u.Hz, equivalencies=u.dimensionless_angles())
+gyroperiod = (1 / freq).to(u.s)
 steps_to_gyroperiod = 10
 timestep = gyroperiod / steps_to_gyroperiod
 
 ############################################################
 # Initialize the trajectory calculation.
-
-number_steps = steps_to_gyroperiod * int(2 * np.pi)
-trajectory = ParticleTracker(plasma, 'p', 1, 1, timestep, number_steps)
-
-############################################################
 # We still have to initialize the particle's velocity. We'll limit ourselves to
 # one in the x direction, parallel to the magnetic field B -
 # that way, it won't turn in the z direction.
 
-trajectory.v[0][0] = 1 * (u.m / u.s)
+trajectory = ParticleTracker(
+    plasma, v=u.Quantity([[-1, 0, 0]] * u.m / u.s), particle_type="p"
+)
 
 ############################################################
-# Run the pusher and plot the trajectory versus time.
+# Let's run the pusher and plot the trajectory versus time.
+# We'll just show the y-z trajectories for clarity.
 
-trajectory.run()
-trajectory.plot_time_trajectories()
+solution = trajectory.run(50 * gyroperiod, timestep)
+solution.particletracker.plot_time_trajectories("yz")
 
 ############################################################
 # Plot the shape of the trajectory in 3D.
 
-trajectory.plot_trajectories()
+solution.particletracker.plot_trajectories()
+
+############################################################
+# If you have Pyvista, you can  run the following code - it'll open up a neat 3D visualization.
+
+# import pyvista
+
+# fig = pyvista.Plotter()
+# solution.visualize(fig)
+# fig.show()
 
 ############################################################
 # As a test, we calculate the mean velocity in the z direction from the
 # velocity and position
 
-vmean = trajectory.velocity_history[:, :, 2].mean()
-print(f"The calculated drift velocity is {vmean:.4f} to compare with the"
-      f"theoretical E0/B0 = {E0/B0:.4f}")
+vmean = solution.velocity.sel(dimension="z").mean().item()
+print(
+    f"The calculated drift velocity is {vmean:.4f} to compare with the "
```

Comment:
Huh!  I didn't know we could do something like f"{vmean:.4f}".  Format codes remind me of the good ole days of Fortran. :smile_cat:

Reply:
You might like https://pyformat.info/ then :)